expr/limit.go

```diff
@@ -72,24 +72,16 @@ type Limit struct {
 }
 
 func (l *Limit) marshal(fam byte) ([]byte, error) {
+	var flags uint32
+	if l.Over {
+		flags = unix.NFT_LIMIT_F_INV
+	}
 	attrs := []netlink.Attribute{
-		{Type: unix.NFTA_LIMIT_TYPE, Data: binaryutil.BigEndian.PutUint32(uint32(l.Type))},
 		{Type: unix.NFTA_LIMIT_RATE, Data: binaryutil.BigEndian.PutUint64(l.Rate)},
 		{Type: unix.NFTA_LIMIT_UNIT, Data: binaryutil.BigEndian.PutUint64(uint64(l.Unit))},
-	}
-
-	if l.Over {
-		attrs = append(attrs, netlink.Attribute{
-			Type: unix.NFTA_LIMIT_FLAGS,
-			Data: binaryutil.BigEndian.PutUint32(unix.NFT_LIMIT_F_INV),
-		})
-	}
-
-	if l.Burst != 0 {
-		attrs = append(attrs, netlink.Attribute{
-			Type: unix.NFTA_LIMIT_BURST,
-			Data: binaryutil.BigEndian.PutUint32(l.Burst),
-		})
+		{Type: unix.NFTA_LIMIT_BURST, Data: binaryutil.BigEndian.PutUint32(l.Burst)},
```

Comment:
Why is the l.Burst handling no longer conditional?

Reply:
I observed that burst is passed with nft regardless of whether it is specified or not:
```
# nft --debug=netlink add rule ip filter forward add @myMeter { ip saddr ct count over 20 limit rate over 10/second } drop
ip filter forward
  [ payload load 4b @ network header + 12 => reg 1 ]
  [ dynset add reg_key 1 set myMeter expr [ connlimit count 20 flags 1 ] expr [ limit rate 10/second burst 5 type packets flags 0x1 ] ]
  [ immediate reg 0 drop ]
```

The same is observed from the raw bytes when burst is not specified in the nft command:
```
# strace -f -v -x -s 2048 -eraw=sendto nft add rule ip filter forward add @myMeter { ip saddr limit rate over 10/second } drop
...
sendmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=[[{nlmsg_len=20, nlmsg_type=NFNL_MSG_BATCH_BEGIN, nlmsg_flags=NLM_F_REQUEST, nlmsg_seq=0, nlmsg_pid=0}, {nfgen_family=AF_UNSPEC, version=NFNETLINK_V0, res_id=htons(2560)}], [{nlmsg_len=272, nlmsg_type=NFNL_SUBSYS_NFTABLES<<8|NFT_MSG_NEWRULE, nlmsg_flags=NLM_F_REQUEST|NLM_F_CREATE|NLM_F_APPEND, nlmsg_seq=1, nlmsg_pid=0}, {nfgen_family=AF_INET, version=NFNETLINK_V0, res_id=htons(0)}, [[{nla_len=11, nla_type=0x1}, "\x66\x69\x6c\x74\x65\x72\x00"], [{nla_len=12, nla_type=0x2}, "\x66\x6f\x72\x77\x61\x72\x64\x00"], [{nla_len=228, nla_type=NLA_F_NESTED|0x4}, "\x34\x00\x01\x80\x0c\x00\x01\x00\x70\x61\x79\x6c\x6f\x61\x64\x00\x24\x00\x02\x80\x08\x00\x01\x00\x00\x00\x00\x01\x08\x00\x02\x00\x00\x00\x00\x01\x08\x00\x03\x00\x00\x00\x00\x0c\x08\x00\x04\x00\x00\x00\x00\x04\x7c\x00\x01\x80\x0b\x00\x01\x00\x64\x79\x6e\x73\x65\x74\x00\x00\x6c\x00\x02\x80\x08\x00\x04\x00\x00\x00\x00\x01\x08\x00\x03\x00\x00\x00\x00\x00\x0c\x00\x01\x00\x6d\x79\x4d\x65\x74\x65\x72\x00\x08\x00\x02\x00\x00\x00\x00\x01\x44\x00\x07\x80\x0a\x00\x01\x00\x6c\x69\x6d\x69\x74\x00\x00\x00\x34\x00\x02\x80\x0c\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x0a\x0c\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x01\x08\x00\x03\x00\x00\x00\x00\x05\x08\x00\x04\x00\x00\x00\x00\x00\x08\x00\x05\x00\x00\x00\x00\x01\x30\x00\x01\x80\x0e\x00\x01\x00\x69\x6d\x6d\x65\x64\x69\x61\x74\x65\x00\x00\x00\x1c\x00\x02\x80\x08\x00\x01\x00\x00\x00\x00\x00\x10\x00\x02\x80\x0c\x00\x02\x80\x08\x00\x01\x00\x00\x00\x00\x00"]]], [{nlmsg_len=20, nlmsg_type=NFNL_MSG_BATCH_END, nlmsg_flags=NLM_F_REQUEST, nlmsg_seq=2, nlmsg_pid=0}, {nfgen_family=AF_UNSPEC, version=NFNETLINK_V0, res_id=htons(2560)}]], iov_len=312}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 312
...
```

From the above `sendmsg` call observe the interpreted part: 
```
# Denotes the limit expression, message length, etc.
limit\x00\x00\x004\x00\x02\x80
# Following parts follow format: 
#   Length | Limit Field ID | Limit Field Data
# 12 | RATE  | 10
\x0c\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\n
# 12 | UNIT  | 1
\x0c\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x01
#  8 | BURST | 5   <-- here it is
\x08\x00\x03\x00\x00\x00\x00\x05
#  8 | TYPE  | 0 (NFT_LIMIT_PKTS)
\x08\x00\x04\x00\x00\x00\x00\x00
#  8 | FLAGS | 16
\x08\x00\x05\x00\x00\x00\x00x10
```

As far as I see, all of the data is passed to the kernel regardless of whether they are specified or not, so the ifs are not needed - if burst is not specified nftables will set burst to 5. I observe the same behaviour with the following struct:
```
# sed -n '52,69p' main.go 
	&expr.Dynset{
		SetID: set.ID,
		SrcRegKey: 1,
		SetName: set.Name,
		Operation: uint32(unix.NFT_DYNSET_OP_ADD),
		Exprs: []expr.Any{
			&expr.Connlimit{
				Count: 20,
				Flags: 1,
			},
			&expr.Limit{
				Type: expr.LimitTypePkts,
				Rate: 10,
				Unit: expr.LimitTimeSecond,
				Over: true,
			},
		},
	},
# go run main.go
filter forward
EXPRS: &{Count:20 Flags:1}
EXPRS: &{Type:0 Rate:10 Over:true Unit:1 Burst:5}
```

Reply:
Thanks for the detailed reply. It took me a while to find this, but `nft` indeed defaults burst to 5, hidden in the parser: https://git.netfilter.org/nftables/tree/src/parser_bison.y?id=994bf5004b365904029f0fe8c2de587178583712#n3304

The Linux kernel itself also defaults to 5, which is likely why you’re seeing the `Burst:5` field when using the Go nftables package: https://github.com/torvalds/linux/blob/39c3c396f8131f3db454c80e0fcfcdc54ed9ec01/net/netfilter/nft_limit.c#L56